mm2src/coins/utxo/utxo_builder/utxo_arc_builder.rs

```diff
@@ -276,102 +293,308 @@ pub(crate) async fn block_header_utxo_loop<T: UtxoCommonOps>(
+                    continue;
+                },
+                TryToRetrieveHeadersUntilSuccessError::PermanentError { .. } => {
+                    error!("{}", err);
+                    remove_server_and_break_if_no_servers_left!(
+                        client,
+                        server_address,
+                        ticker,
+                        sync_status_loop_handle
+                    );
+                    continue;
+                },
+            },
+        };
+
+        // Validate retrieved block headers.
+        if let Err(err) = validate_headers(ticker, last_height_in_storage, &block_headers, storage, &spv_conf).await {
+            error!("Error {} on validating the latest headers for {}!", err, ticker);
+            // This code block handles a specific error scenario where a parent hash mismatch(chain re-org) is
+            // detected in the SPV client.
+            // If this error occurs, the code retrieves and revalidates the mismatching header from the SPV client..
+            if let SPVError::ParentHashMismatch {
+                coin,
+                mismatched_block_height,
+            } = &err
+            {
+                match resolve_possible_chain_reorg(
+                    client,
+                    server_address,
+                    &mut args,
+                    last_height_in_storage,
+                    *mismatched_block_height,
+                    storage,
+                    &spv_conf,
+                )
+                .await
+                {
+                    Ok(()) => {
+                        info!(
+                            "Chain reorg detected and resolved for coin: {}, re-syncing reorganized headers!",
+                            coin
+                        );
+                        continue;
+                    },
+                    Err(err) => {
+                        error!("Error {} on resolving chain reorg for coin: {}!", err, coin);
+                        if err.get_inner().is_network_error() {
+                            sync_status_loop_handle.notify_on_temp_error(err.to_string());
+                        } else {
+                            remove_server_and_break_if_no_servers_left!(
+                                client,
+                                server_address,
+                                ticker,
+                                sync_status_loop_handle
+                            );
+                        }
+                        continue;
+                    },
+                }
+            }
+            remove_server_and_break_if_no_servers_left!(client, server_address, ticker, sync_status_loop_handle);
+            continue;
+        }
+
+        let sleep = args.error_sleep;
+        ok_or_continue_after_sleep!(storage.add_block_headers_to_storage(block_registry).await, sleep);
+    }
+}
+
+#[derive(Debug, Display)]
+enum TryToRetrieveHeadersUntilSuccessError {
```

Comment:
I think this naming is unnecessary too long, we can just have `RetrieveHeadersError`

Reply:
but `retrieve_headers` is another function where this error isn't used and shouldn't be used in the future, that's why I need a different name.
https://github.com/KomodoPlatform/atomicDEX-API/blob/ab1dd8fc598562b5f78dc8728fb8367574f1596e/mm2src/coins/utxo/rpc_clients.rs#L2095
I don't see any reason to not have a long name if it describes the error accurately. If for example I used `RetrieveHeadersUntilSuccessError` as a name instead, the name is kind of contradictory, that's why I added `TryTo` to the beginning .